spec.bs

```diff
@@ -1844,11 +1867,26 @@ Expand the section of the navigation/traversal response handling which deals wit
 <h3 id="navigate-event-traversal-patches">History traversal updates</h3>
 
 <div algorithm="apply the history step">
-  Modify the [=apply the history step=] algorithm as follows. Near the end of the algorithm, inside the queued task over each global (currently step 14.10), before potentially unloading or activating the history entry, add the following step:
+  Modify the [=apply the history step=] algorithm as follows. Near the start of the algorithm, before checking if unloading is user-canceled (currently step 5), add the following steps:
+  1. Let <i>targetEntry</i> be the result of [=getting the target history entry=] given <i>traversable</i> and <i>targetStep.</i>
+  1. If <i>targetEntry</i> does not equal <i>traversable</i>'s [=navigable/current session history entry=], then:
+    1. Let <i>continue</i> be the result of [=Firing a traversal navigate event=] at <i>traversable</i>'s [=navigable/active window=]'s [=Window/navigation API=] with <i>[=fire a traversal navigate event/destinationEntry=]</i> set to <i>targetEntry</i> and <i>[=fire a traversal navigate event/userInvolvement=]</i> set to <var ignore>userInvolvement</var>
+    1. If <i>continue</i> is false, abort these steps.
 
-  1. [=Fire a traversal navigate event=] at <var ignore>displayedDocument</var>'s [=relevant global object=]'s [=Window/navigation API=] with <i>[=fire a traversal navigate event/destinationEntry=]</i> set to <var ignore>targetEntry</var> and <i>[=fire a traversal navigate event/userInvolvement=]</i> set to <var ignore>userInvolvement</var>.
+  After setting navigable's ongoing navigation to "traversal" (currently step 8.3), add the following step:
+  1. If <i>navigable</i> does not equal <i>traversable</i>, then:
+    1. Let <i>continue</i> be the result of [=Firing a traversal navigate event=] at <i>navigable</i>'s [=navigable/active window=]'s [=Window/navigation API=] with <i>[=fire a traversal navigate event/destinationEntry=]</i> set to <i>targetEntry</i> and <i>[=fire a traversal navigate event/userInvolvement=]</i> set to <var ignore>userInvolvement</var>
+    1. [=Assert=]: <i>continue</i> is true.
 
 (Recall that we introduced the <var ignore>userInvolvement</var> parameter as part of [[#user-initiated-patches]].)
+
+  Before the update document for history step application (currently step 14.10.3), add the following step:
+  1. If <i>targetEntry</i>'s [=document=] is equal to <i>displayedDocument</i>:
+    1. Let <i>updateDocument</i> be the result of [=maybe continue the navigate event=] given <i>displayedDocument</i>'s [=relevant global object=]'s [=Window/navigation API=].
+    1. If <i>updateDocument</i> is false:
```

Comment:
I think this will never be false? For the traverse case, "react to navigate event result" always returns true. So "maybe continue the navigate event" probably doesn't need a return value? And we shouldn't skip any jobs.

Basically, the return value of "react to navigate event result" is used for determining whether to stop a cross-document non-traversal nav. For the traversal case, we only allow intercepting if it's a same-document traversal anyway, so we should just let the normal same-document traversal stuff happen.

Reply:
It's possible I've misunderstood, but in the traversal case, I thought we also needed to ensure that "update document for history application step" does *not* run when the traversal was intercepted. That was my goal in reading the return value here. I've update the return value of "react to navigate event result" to essentially mean "was not intercepted" to ensure we're debating what I actually meant to write. :)

Reply:
I think this is maybe an intentional departure between the implementation and spec. In the spec, same-document traversals always go down the "normal path" (i.e., "update document for history application step"), and never do the URL and history update steps. Whereas in the implementation, IIRC, you have some version of the URL and history update steps which works for traversals too, and you do that instead.